test/Nerdbank.MessagePack.Tests/KnownSubTypeTests.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Andrew Arnott. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+public partial class UnionTests(ITestOutputHelper logger) : MessagePackSerializerTestBase(logger)
+{
+	[Fact]
+	public void BaseType()
+	{
+		ReadOnlySequence<byte> msgpack = this.AssertRoundtrip(new BaseClass { BaseClassProperty = 5 });
+
+		// Assert that it's serialized in its special syntax that allows for derived types.
+		MessagePackReader reader = new(msgpack);
+		Assert.Equal(2, reader.ReadArrayHeader());
+		reader.ReadNil();
+		Assert.Equal(1, reader.ReadMapHeader());
+		Assert.Equal(nameof(BaseClass.BaseClassProperty), reader.ReadString());
+	}
+
+	[Fact]
+	public void DerivedAType()
+	{
+		ReadOnlySequence<byte> msgpack = this.AssertRoundtrip(new DerivedA { BaseClassProperty = 5, DerivedAProperty = 6 });
+
+		// Assert that this has no special header because it has no Union attribute of its own.
+		MessagePackReader reader = new(msgpack);
+		Assert.Equal(2, reader.ReadMapHeader());
+		Assert.Equal(nameof(DerivedA.DerivedAProperty), reader.ReadString());
+		Assert.Equal(6, reader.ReadInt32());
+		Assert.Equal(nameof(BaseClass.BaseClassProperty), reader.ReadString());
+		Assert.Equal(5, reader.ReadInt32());
+		Assert.True(reader.End);
+	}
+
+	[Fact]
+	public void DerivedA_AsBaseType() => this.AssertRoundtrip<BaseClass>(new DerivedA { BaseClassProperty = 5, DerivedAProperty = 6 });
+
+	[Fact]
+	public void DerivedAA_AsBaseType() => this.AssertRoundtrip<BaseClass>(new DerivedAA { BaseClassProperty = 5, DerivedAProperty = 6 });
+
+	[Fact]
+	public void DerivedB_AsBaseType() => this.AssertRoundtrip<BaseClass>(new DerivedB(10) { BaseClassProperty = 5 });
+
+	[GenerateShape]
+	[KnownSubType(1, typeof(DerivedA))]
```

Comment:
Perhaps you could introduce a degree of type safety here by making the attribute generic:
```csharp
public class KnownSubTypeAttribute<TSubType>(int tag) where TSubType : IShapeable<TSubType>
```

Reply:
The downside is that it forces users to apply the source generator.

Reply:
I'm planning on writing analyzers to apply constraints to this type, including ensuring that it appears only once and derives from the type the attribute is applied to.
I'm divided on whether to require the source generator. I've recently added Serialize method overloads that allow providing a shape as an ordinary argument, so I'm kinda moving in the direction of supporting runtime supplied shapes. But not sure yet... we'll see.